Implement functions to add/set unencoded path attributes

stefanberger opened the pull request.

stefanberger committed:
Implement functions to add/set unencoded path attributes

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

dwmw2 commented:
Works for me; thanks.

Might we also want to set query attributes unencoded? If we've manually prompted the user for a password/PIN, I can imagine wanting to then add that as `pin-value` in the internally stored URI? And won't we want to add that as an unescaped `[]byte` too?

stefanberger committed:
Implement functions to add/set unencoded query attributes

Adjust the tests cases to also test unencoded query attributes. Use an
already-escaped PIN value in the existing tests to check that GetPIN()
returns the unescaped string.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>

stefanberger commented:
> Might we also want to set query attributes unencoded? If we've manually prompted the user for a password/PIN, I can imagine wanting to then add that as `pin-value` in the internally stored URI? And won't we want to add that as an unescaped `[]byte` too?

Yes, I added 2 functions now for adding unencoded query attributes.

dwmw2 commented on pkcs11uri_test.go:
```diff
@@ -78,27 +78,73 @@ func verifyPIN(t *testing.T, uri *Pkcs11URI, expectedpin string) {
+func TestConstruct4(t *testing.T) {
+	uri := New()
+
+	err := uri.AddPathAttributeUnencoded("slot-id", []byte("12345"))
```
Bit of a boring test. The encoding on "12345" (`0x31 0x32 0x33 0x34 0x45`) does nothing, which I presume is why the test succeeded even when you were double-encoding? 

It might be better to test with actual percent signs in the values, like `id=%25%30%31` (three bytes `0x25 0x30 0x31` or the literal string `%01`). And with non-ASCII characters that *need* encoding, such as `id=%00` which is `[]byte{0}`.

stefanberger replied:
> Bit of a boring test. The encoding on "12345" (`0x31 0x32 0x33 0x34 0x45`) does nothing, which I presume is why the test succeeded even when you were double-encoding?

This was an existing test and slot-id **must** be an integer...

The test that had passed was calling `uri.AddPathAttributeUnencoded("id", []byte{0x66,0x6f,0x6f})` and did match `"pkcs11:id=%66%6F%6F"` without causing double-quoting.

stefanberger replied:
I modified the test now to use `the pin` as PIN.

stefanberger replied:
Unless there are other concerns I will merge this patch.

stefanberger replied:
Anyone got any comments on this patch? Otherwise I'd merge it soon.
